Social: Enforce sharing limits in the block editor

danielpost opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

This PR enforces sharing limits in the block editor, if applicable.

There's an edge case that makes it possible to bypass the limit by adding a new connection during the publishing process—we can fix this in a follow-up PR.

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* Check if the sharing limit is enabled for a specific site.
* If it is, ensure users cannot enable more connections than they have shares left.
* Disable connections by default in the block editor when the number of connections is greater than the number of shares left.

#### Other information:

- [X] Have you written new tests for your changes, if applicable?
- [X] Have you checked the E2E test CI results, and verified that your changes do not break them?

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
pdrWKz-wQ-p2
p6TEKc-6td-p2

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No.

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

_Please note that this PR needs D85993-code to be merged. In the meantime, you can sandbox it to test._

* Make sure tests pass: `jetpack test plugins/social php`
* Enable the Jetpack Social plugin.
* Make a new post, and check that the sharing screen looks like the first image.
* Enable sharing limits for your site (first line in test plan of D85993-code)
* Make a new post, and check that you can see the upgrade notice (second image).
* Test that connections are disabled by default when there's more connections than shares left, and that you can't select more connections than allowed.
  * To mock the number of shares left, you can add `$shares['shares_remaining'] = 1;` to `initial_state()` and `$shares_remaining = 1;` to `set_up_sharing_limits()` in `jetpack-social.php`
* Finally, enable the Jetpack plugin, and make sure the screen looks like the first screenshot again

<table>
<tr>
	<td>Sharing Limits Disabled 
	<td>Sharing Limits Enabled
	<td>More connections than shares left
	<td>More connections than shares left, and all remaining shares have been used
<tr>
	<td><img width="278" alt="image" src="https://user-images.githubusercontent.com/1713699/184976111-03be1d83-04b1-449c-bdea-583b8281a292.png">
	<td><img width="278" alt="image" src="https://user-images.githubusercontent.com/1713699/184976450-e0c32894-9eb3-4402-a5ea-826f367c2d25.png">
	<td><img width="280" alt="image" src="https://user-images.githubusercontent.com/1713699/184976463-949ade72-0653-4c15-95a3-e20fa62945a4.png">
	<td><img width="280" alt="image" src="https://user-images.githubusercontent.com/1713699/184976475-d15d5279-9561-4d0f-b536-f7b34e431524.png">
</table>

Commit:
Enforce sharing limits

Commit:
Changelog

Commit:
Simplify filter

Commit:
Add tests

Commit:
Fix missing number

spsiddarthan commented on projects/js-packages/publicize-components/src/components/form/index.js:
```diff
@@ -22,41 +24,85 @@ import styles from './styles.module.scss';
 	const Wrapper = isPublicizeDisabledBySitePlan ? Disabled : Fragment;
 
+	const outOfConnections =
+		numberOfSharesRemaining !== null && numberOfSharesRemaining <= enabledConnections.length;
```
Can we make this `const outOfConnections = numberOfSharesRemaining && numberOfSharesRemaining <= enabledConnections.length;`

danielpost replied:
Unfortunately not, because `0` is considered a falsy value, which means that `outOfConnections` would always return false if the number of shares remaining equals `0`.

Personally, I don't mind the explicit `!== null` check too much here, because it makes it clear that a value has been passed to the `numberOfSharesRemaining` prop.

spsiddarthan commented on projects/js-packages/publicize-components/src/components/form/index.js:
```diff
@@ -22,41 +24,85 @@ import styles from './styles.module.scss';
-					</ul>
-				</PanelRow>
+				<>
+					{ numberOfSharesRemaining !== null && (
```
May be we can do `numberOfSharesRemaining &&  ( ` here?

danielpost replied:
Same as above ^

spsiddarthan commented on projects/js-packages/publicize-components/src/components/form/index.js:
```diff
@@ -22,41 +24,85 @@ import styles from './styles.module.scss';
+										{
+											upgradeLink: (
+												<a
+													href={ 'https://example.com' }
```
Should we add the link to check out the plan here?

danielpost replied:
I wasn't sure if we should add it yet, considering it's not really publicly available yet. What do you think?

spsiddarthan replied:
Makes sense, we can replace the link when we are ready to launch

spsiddarthan commented on projects/plugins/social/src/js/components/panel/index.jsx:
```diff
@@ -71,6 +83,7 @@ const PublicizePanel = ( { prePublish } ) => {
 				<PublicizeForm
 					isPublicizeEnabled={ isPublicizeEnabled }
 					isRePublicizeFeatureEnabled={ ! isPostPublished }
+					numberOfSharesRemaining={ isShareLimitEnabled ? numberOfSharesRemaining : undefined }
```
Nitpick: Should we pass`null` instead of `undefined`?

danielpost replied:
Yes, good idea. It'll be consistent with the default value in the `PublicizeForm` component.

spsiddarthan commented on projects/plugins/social/tests/php/test-class-share-limits.php:
```diff
@@ -0,0 +1,50 @@
+/**
+ * Main plugin file testing.
+ */
+class Share_Limits_Test extends BaseTestCase {
```
Great job adding the test :)

danielpost replied:
Thanks! 😄

spsiddarthan commented on projects/js-packages/publicize-components/src/components/form/index.js:
```diff
@@ -22,41 +24,85 @@ import styles from './styles.module.scss';
+				<>
+					{ numberOfSharesRemaining !== null && (
+						<PanelRow>
+							<Notice type={ numberOfSharesRemaining < connections.length ? 'warning' : 'default' }>
```
I know we spoke about this on slack, but I was thinking some more about this. I reckon we should show the notice only when the user doesn't have the jetpack social plan. That's controlled not just by the sticker, but also by the presence of a plan. 

I had a Jetpack Social Basic Monthly plan and could still see the notice. 

So, may be show the notice only if the user doesn't have the paid plan? We can use [this flag ](https://github.com/Automattic/jetpack/blob/fe38d3671f2db213277bd1ea6d1b819b780561a0/projects/plugins/social/src/class-jetpack-social.php#L153:~:text=%24show_nudge%20%20%20%20%20%20%20%20%20%20%20%20%20%20%3D%20!%20Current_Plan%3A%3Asupports(%20%27social%2Dshares%2D1000%27%2C%20%24refresh_plan_from_wpcom%20)%3B) to determine that.

danielpost replied:
Yes, good idea. I'll implement that.

pablinos replied:
Has that been implemented? Is that wrapped up in the `isDisabled` check?

danielpost replied:
It's implemented [here](https://github.com/Automattic/jetpack/pull/25661/files#diff-58f304088488b354751173257592c43671853b5b733dbdc410f057ea0b1c6bf4R88), by only passing in `numberOfSharesRemaining` if the user doesn't have a paid plan.

spsiddarthan reviewed (changes requested):
Thanks for your work on this! 

All your test instructions work perfectly. I have added a few minor comments and made a remark about not showing the Notice component when we have a plan. 

Great work overall!

spsiddarthan commented on projects/plugins/social/src/class-jetpack-social.php:
```diff
@@ -279,4 +283,27 @@ public function calculate_scheduled_shares() {
+		}
+
+		$connections      = $publicize->get_all_connections();
+		$shares_remaining = 1; // @TODO: add this calculated number to the endpoint.
```
Just confirming that this is hard-coded just for testing purposes.

danielpost replied:
No, I forgot to push a change 😂 will fix!

Commit:
Switch undefined to null for consistency

Commit:
Merge branch 'trunk' into change/enforce-sharing-limit-in-block-editor

Commit:
Do not show notice for users with paid plans

Commit:
Fix dependencies

gmjuhasz commented:
LGTM, I agree with the changes and keeping the null check! What do you think @spsiddarthan?

Commit:
Only show notice and enforce limits for users without paid plan

spsiddarthan reviewed (dismissed):
Good work, @danielpost! Do we need crew's review on this one or can we merge on our one? 

We will need to resolve the conflicts though. Also, I would recommend adding this to the list of things to be beta tested (since we are tweaking a core functionality on Jetpack and of course the testing instructions for those can be a lot simpler - test publicizing on the Jetpack plugin and make sure sharing to your connections is enabled by default)

danielpost commented on pnpm-lock.yaml:
```diff
@@ -1762,6 +1762,9 @@ importers:
       config: 3.3.7
       jetpack-e2e-commons: link:../../../../../tools/e2e-commons
 
+  projects/plugins/super-cache:
+    specifiers: {}
+
```
@Automattic/jetpack-crew Is this change expected? I'm not sure where it came from.

anomiex replied:
It's harmless.

`projects/plugins/super-cache` was recently added, but since it has no JS dependencies yet it's optional and so the "lock file is up to date" check didn't flag it as missing. When you did something here that updated the lock file, it got added then.

And then when you did the trunk merge in c3295ff, that dropped the addition from your PR because #25687 did the addition along with its own changes.

Commit:
Remove unnecessary Fragment component

Commit:
Merge branch 'trunk' into change/enforce-sharing-limit-in-block-editor

jeherve commented on projects/js-packages/publicize-components/src/components/form/index.js:
```diff
@@ -22,41 +24,79 @@ import styles from './styles.module.scss';
+									),
+									{
+										upgradeLink: (
+											<a href={ 'https://example.com' } rel="noopener noreferrer" target="_blank" />
```
Maybe we could use `ExternalLink` here (and maybe also have a to-do comment so we do not forget to update that link)?

danielpost replied:
Sure, I'll give it a shot!

danielpost replied:
Fixed in [7faf0fa](https://github.com/Automattic/jetpack/pull/25661/commits/7faf0faa596ed4625361a78e310648b834b61792).

jeherve commented on projects/js-packages/publicize-components/src/components/notice/styles.module.scss:
```diff
@@ -0,0 +1,15 @@
+.notice {
+	color: #757575;
+	margin-bottom: 16px;
+}
+
+.notice--warning {
+	padding: 8px 16px;
+	color: #4f3500;
+	background-color: #f5f1e1;
+	border-left: 2px solid #f0c930;
```
Can we leverage `@automattic/jetpack-base-styles` for some of the values here?

danielpost replied:
Is that possible? I wasn't sure to what extent we could use our existing components in the Gutenberg editor. Perhaps I can wrap it in a [`ThemeProvider`](https://github.com/Automattic/jetpack/tree/trunk/projects/js-packages/components/components/theme-provider) and get the styles like that.

jeherve replied:
👍 You can check how we use this in other blocks for some examples.

danielpost replied:
Fixed in [c424135](https://github.com/Automattic/jetpack/pull/25661/commits/c424135adb3b9052e4e61183b3ad580254ad7f7b).

jeherve reviewed (commented):
I did not fully test this, but wanted to make a couple of minor suggestions.

coder-karen commented on projects/plugins/social/src/class-share-limits.php:
```diff
@@ -0,0 +1,58 @@
+			return;
+		}
+
+		add_filter( 'publicize_checkbox_default', '__return_false' );
```
When I set the sharing limit to 1 with 2 connected accounts, the post was still shared despite the toggles for sharing being off because of the addition of this filter. I suspect that is due to the issue that this PR set out to fix: https://github.com/Automattic/jetpack/pull/24617

I imagine that would cause some confusion unless the intention is to fix that before this goes live too?

danielpost replied:
Ah good catch! Yes, that definitely needs to be merged before we enable this for all users.

pablinos replied:
We've just merged #24617, so this should work as expected now.

coder-karen replied:
Ah nice. I notice I still have the same issue (despite the toggles being off for each account, the post was still shared - to both, with a sharing limit of 1 in place) though - I'm testing in a local dev environment and tried re-installing JP social and rebuilding it, making sure the branch is up-to-date. Not too sure what I'm missing here.

coder-karen reviewed (changes requested):
Nice feature, the notice works well. I have left one comment below about the `publicize_checkbox_default` filter.

jeherve reviewed (commented):
The UI seems to be displayed for already published posts as well; I don't think it should be?

<img width="302" alt="image" src="https://user-images.githubusercontent.com/426388/186728837-ed822be3-4fa0-49dc-a013-85a04c6ea4d3.png">

*********

I tested with setting the shares remaining to 0, and the UI just disappeared without stopping me from publicizing. Is that expected right now?

Commit:
Merge branch 'trunk' into change/enforce-sharing-limit-in-block-editor

Commit:
Use ExternalLink and Jetpack Redirect for upgrade nudge

Commit:
Use existing variables for notice styling

Commit:
Do not show notice when Jetpack Social is disabled

danielpost commented:
@jeherve I addressed your feedback, thanks!

> I tested with setting the shares remaining to 0, and the UI just disappeared without stopping me from publicizing. Is that expected right now?

How and where did you set it to 0? It should look like this:

<img width="281" alt="image" src="https://user-images.githubusercontent.com/1713699/186875261-d916a1dd-450a-40af-a9d8-d3f8e5f80021.png">

jeherve commented:
> How and where did you set it to 0? It should look like this:

I set it in the same place as suggested in the PR description (line 154 and 307). The warning disappears entirely for me.

gmjuhasz commented on projects/js-packages/publicize-components/src/components/notice/index.js:
```diff
@@ -0,0 +1,16 @@
+
+Notice.propTypes = {
+	children: PropTypes.node.isRequired,
+	type: PropTypes.oneOf( [ 'default', 'warning' ] ),
```
I think this should include _error_ as well, isn't it?

danielpost replied:
Yes thanks, I forgot to update that when I added the new type.

Commit:
Fix logic in selectors

danielpost commented:
@jeherve Good catch! There was a mistake in the store selectors that caused the default value to be used for any false-y values, which in the case of `numberOfSharesRemaining` turned `0` into `null`. Fixed it 👍🏼

Commit:
Add missing type

pablinos commented on projects/js-packages/publicize-components/src/components/form/index.js:
```diff
@@ -22,41 +25,81 @@ import styles from './styles.module.scss';
+								( { display_name, enabled, id, service_name, toggleable, profile_picture } ) => (
+									<PublicizeConnection
+										disabled={
+											( isRePublicizeFeatureEnabled ? ! isPublicizeEnabled : ! toggleable ) ||
```
This looks like the plan checks that we want to remove for Republicize. Should we do that now?

danielpost replied:
@spsiddarthan is working on a PR to remove these, I think it would be good not to touch that logic in this PR and instead bring them together later.

pablinos replied:
Ah of course. I forgot that these checks were actually being used currently!

pablinos commented on projects/js-packages/publicize-components/src/components/form/index.js:
```diff
@@ -22,41 +25,81 @@ import styles from './styles.module.scss';
+									<PublicizeConnection
+										disabled={
+											( isRePublicizeFeatureEnabled ? ! isPublicizeEnabled : ! toggleable ) ||
+											( ! enabled && toggleable && outOfConnections )
```
I wonder if we should put this logic into another helper function, perhaps in the hook, or inside the `PublicizeConnection` component.

danielpost replied:
Yeah we can do that in a follow-up PR. I'm curious to see what the logic will look like without the Republicize checks.

pablinos commented on projects/js-packages/publicize-components/src/components/form/index.js:
```diff
@@ -22,41 +25,81 @@ import styles from './styles.module.scss';
+									sprintf(
+										/* translators: %d is the number of shares remaining, upgradeLink is the link to upgrade to a different plan */
+										_n(
+											'You have %d share remaining. <upgradeLink>Upgrade</upgradeLink> to share to all your social media accounts.',
```
Maybe for a follow up change but when I have more shares remaining than connections, the phrase 'to share to all your social media accounts' doesn't make as much sense because the post will be shared to all accounts. It should be 'to share this post as many times as you like' or 'to remove this limit'.

It's probably on the edge at the moment, but with this logic, should we make this a separate `ShareLimitNotice` component?

danielpost replied:
Yeah, we can create a new component in a follow-up PR. Good idea 👍🏼

pablinos commented on projects/plugins/social/src/class-jetpack-social.php:
```diff
@@ -203,6 +206,8 @@ class_exists( 'Jetpack' ) ||
 
 		Assets::enqueue_script( 'jetpack-social-editor' );
 
+		wp_add_inline_script( 'jetpack-social-editor', $this->render_initial_state(), 'before' );
```
Should this be added into (or replace) the `wp_localize_script` call below?

danielpost replied:
We have shared state for both the editor and admin page, and then some variables unique to each. I can make a follow-up PR to improve how this works.

pablinos replied:
Yeah it seems like we want all the initial state now for both the editor and the admin page. Perhaps we should just merge the arrays for now?

pablinos replied:
Looking at this a bit more, I think longer term we should have some sort of `initial_state` class that can be used for rendering this data. I notice that there's some for the Connection package for example.

pablinos commented on projects/plugins/social/src/class-jetpack-social.php:
```diff
@@ -101,6 +101,8 @@ function () {
 
 		add_filter( 'jetpack_get_available_standalone_modules', array( $this, 'social_filter_available_modules' ), 10, 1 );
+
+		add_action( 'admin_init', array( $this, 'set_up_sharing_limits' ) );
```
I wonder if we could hook this into specific screens rather than every admin call. It probably doesn't matter on the average site, but could cause a problem on large multisites like WPCOM

danielpost replied:
Good idea, I'll look into this.

pablinos reviewed (commented):
A couple of thoughts looking through the code, and from the position of not having been a part of the development of this, so feel free to ignore it all!

Commit:
Merge branch 'trunk' into change/enforce-sharing-limit-in-block-editor

Commit:
Merge branch 'trunk' into change/enforce-sharing-limit-in-block-editor

pablinos reviewed (approved):
This is working for me. It's still behind the feature flag, so I think we can merge it and iterate on it in subsequent PRs.

bindlegirl reviewed (approved):
Tests worked well for me.

coder-karen commented:
Just adding here that I added another comment [in this thread](https://github.com/Automattic/jetpack/pull/25661#discussion_r957280384) (despite the toggles being off for each account, the post was still shared - to both, with a sharing limit of 1 in place)

coder-karen committed:
Merge branch 'trunk' into change/enforce-sharing-limit-in-block-editor

coder-karen reviewed (approved):
Approving, based on p1661786884729909-slack-CBG1CP4EN

danielpost commented:
r251507-wpcom
